packages/next/src/client/flight-data-helpers.ts

```diff
@@ -0,0 +1,59 @@
+import type {
+  CacheNodeSeedData,
+  FlightDataPath,
+  FlightRouterState,
+  FlightSegmentPath,
+  Segment,
+} from '../server/app-render/types'
+
+export function getFlightDataPartsFromPath(flightDataPath: FlightDataPath): {
+  /**
+   * The full `FlightSegmentPath` inclusive of the final `Segment`
+   */
+  segmentPath: FlightSegmentPath
+  /**
+   * The `FlightSegmentPath` exclusive of the final `Segment`
+   */
+  pathToSegment: FlightSegmentPath
+  segment: Segment
+  tree: FlightRouterState
+  seedData: CacheNodeSeedData | null
+  head: React.ReactNode | null
+} {
+  // tree, seedData, and head are *always* the last three items in the `FlightDataPath`.
+  const [tree, seedData, head] = flightDataPath.slice(-3)
+  // The `FlightSegmentPath` is everything except the last three items. For a root render, it won't be present.
+  const segmentPath = flightDataPath.slice(0, -3)
+
+  return {
+    // TODO: Unify these two segment path helpers. We are inconsistently pushing an empty segment ("")
+    // to the start of the segment path in some places which makes it hard to use solely the segment path.
+    // Look for "// TODO-APP: remove ''" in the codebase.
+    pathToSegment: segmentPath.slice(0, -1),
+    segmentPath,
+    // if the `FlightDataPath` corresponds with the root, there'll be no segment path,
+    // in which case we default to ''.
+    segment: segmentPath[segmentPath.length - 1] ?? '',
+    tree,
+    seedData,
+    head,
+  }
```

Comment:
If these become perf critical, we should probably split this up into separate functions since not all callsites use all parts